### PR TITLE
(SIMP 7227) Add nmcli support for DNS servers

### PR DESCRIPTION
**Pull request by kendall-moore**
- Allows users to update IPv4 DNS servers via nmcli
- Allows user to specify speciic device for DNS servers

[SIMP-7227] #close

[SIMP-7227]: https://simp-project.atlassian.net/browse/SIMP-7227

### CODE REVIEW

**Comment by trevor-vaughan** [comment]
@op-ct the `simplib__networkmanager` fact was created so that you could prevent `nmcli` from running wild on other parts of the system. If we're 100% certain (and I think we are) that mucking about with the resolv bits of `nmcli` won't affect the rest of the system, then I think the way this is implemented is correct.

I think that we should be able to test this via acceptance tests. The `network` tests were difficult because `nmcli` activating would destroy the Vagrant-specific networking. However, if this works properly, then we should just be messing about with `resolv` settings and this should not break the Vagrant hooks.

**Comment by kendall-moore** [comment]
@trevor-vaughan @op-ct I've been having a bear of a time getting the simp beaker tests to run locally, but I'm hoping that I can work through that this afternoon. I'll try to add a Beaker test accordingly today as well, but I'll update you on progress. I believe the way `nmcli` is used in this context shouldn't impact anything other than `resolv.conf` but I guess I won't know that for sure until later.

File: manifests/init.pp [hunk 1]
```diff
@@ -40,30 +40,68 @@
 #   *If* the $servers array above starts with '127.0.0.1' or '::1', then
 #   the system will set itself up as a caching nameserver unless this is set
 #   to false.
+# @param manage_via_nmcli
```
Comment (trevor-vaughan) [inline manifests/init.pp:43]: @kendall-moore  Can we change this to `use_nmcli` to match the majority of the other modules? Not critical, just nice to have.

File: manifests/init.pp [hunk 2]
```diff
@@ -40,30 +40,68 @@
...
+  Boolean                    $inet6                     = false,
+  Integer[0,15]              $ndots                     = 1,
+  Integer[0,30]              $timeout                   = 2,
+  Integer[0,5]               $attempts                  = 2,
+  Boolean                    $named_server              = false,
+  Boolean                    $named_autoconf            = true,
+  Boolean                    $caching                   = true,
+  Boolean                    $manage_via_nmcli          = false,
+  Optional[String]           $nmcli_device_name         = undef,
+  Boolean                    $auto_reapply_nmcli_device = false,
```
Comment (trevor-vaughan) [inline manifests/init.pp:74]: Not sure why you need this. The DNS settings should be global (I think).

A reapply may trigger unintended updates if applied to a given device where alternative management is in use.

Comment (trevor-vaughan) [reply manifests/init.pp:74]: Actually, I just figured it out and it's a bit unfortunate.

So, in the near future, we're probably going to run into issues with wanting different DNS servers per interface and needing to hook into `systemd-resolved`.

For now, you probably want to auto-detect the first active nmcli device since this updates global settings anyway.

The `simplib__networkmanager` fact should help with this. That said, if the interface isn't under NetworkManager control I have no idea what will happen in general.

File: manifests/init.pp (hunk 1 above)
Comment (kendall-moore) [reply manifests/init.pp:43]: Sure, didn't see this but let me change it now and push.

**Comment by trevor-vaughan** [comment]
@kendall-moore Can you add the Beaker tests to the `.gitlab-ci.yml` and push it back over to GitLab for testing?

You can find an example at https://github.com/simp/pupmod-simp-compliance_markup/blob/master/.gitlab-ci.yml#L160-L192

**Comment by kendall-moore** [comment]
@trevor-vaughan sure: https://gitlab.com/simp/pupmod-simp-resolv/-/pipelines/157364258

**Comment by trevor-vaughan** [comment]
Checking a couple of edge cases and re-prodding the acceptance test over at https://gitlab.com/simp/pupmod-simp-resolv/-/pipelines/157435562

**Comment by trevor-vaughan** [comment]
@op-ct I think that this failure is an issue with the GitLab runners https://gitlab.com/simp/pupmod-simp-resolv/-/jobs/601335826.

Probably need to add the following to the running user's `~/.ssh/config`

```
KexAlgorithms +ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521
MACs +hmac-sha1,hmac-sha2-256
```

**Comment by op-ct** [comment]
@trevor-vaughan Thanks!  I'll take a look.

**Comment by trevor-vaughan** [comment]
Working on a repeater in the acceptance tests

**Comment by trevor-vaughan** [comment]
Ah, wait, I see that this was expected. OK, this is fine.

**Comment by trevor-vaughan** [comment]
Merging even with the failing GitLab test since that seems to be a runner error.

**Comment by op-ct** [comment]
> @op-ct I think that this failure is an issue with the GitLab runners https://gitlab.com/simp/pupmod-simp-resolv/-/jobs/601335826.
> 
> Probably need to add the following to the running user's `~/.ssh/config`
> 
> ```
> KexAlgorithms +ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521
> MACs +hmac-sha1,hmac-sha2-256
> ```

An update to the failing FIPS test:
* The missing SSH setting is `Ciphers: +aes256-cbc,aes192-cbc,aes128-cbc`/
* beaker-vagrant needs a tweak in order to reliably configure SUTs to use it
* Tracking in issue [SIMP-7825][0].

[0]: https://simp-project.atlassian.net/browse/SIMP-7825?focusedCommentId=39084&oldIssueView=true&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-39084
